### ISSUE TITLE
Not Found Error in GitHub publish step

### ISSUE THREAD

**Issue by felixfbecker** [semantic-release/semantic-release#607]
## Current behavior

```
[Semantic release]: Wrote version 1.5.0 to package.json
[Semantic release]: Publishing version 1.5.0 to npm registry
+ chromeless@1.5.0[Semantic release]: An error occurred while running semantic-release: { [Error: {"message":"Not Found","documentation_url":"https://developer.github.com/v3/git/refs/#create-a-reference"}]
  message: '{"message":"Not Found","documentation_url":"https://developer.github.com/v3/git/refs/#create-a-reference"}',
  code: 404,
  status: 'Not Found',
  headers: 
   { server: 'GitHub.com',
     date: 'Thu, 18 Jan 2018 04:52:17 GMT',
     'content-type': 'application/json; charset=utf-8',
     'content-length': '106',
     connection: 'close',
     status: '404 Not Found',
     'x-ratelimit-limit': '5000',
     'x-ratelimit-remaining': '4763',
     'x-ratelimit-reset': '1516253265',
     'x-oauth-scopes': 'read:org, repo:status, repo_deployment, user:email, write:repo_hook',
     'x-accepted-oauth-scopes': '',
     'x-github-media-type': 'github.v3; format=json',
     'access-control-expose-headers': 'ETag, Link, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval',
     'access-control-allow-origin': '*',
     'content-security-policy': 'default-src \'none\'',
     'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
     'x-content-type-options': 'nosniff',
     'x-frame-options': 'deny',
     'x-xss-protection': '1; mode=block',
     'x-runtime-rack': '0.021744',
     'x-github-request-id': 'AC00:7532:CE39AE:188EE14:5A602801' } }
```

## Expected behavior

No error, publish release on GitHub.

## Environment

- **semantic-release** version: 12.2.2, 
- CI environment: CircleCI
- Plugins used: @semantic-release/npm@2.6.1, @semantic-release/github@3.0.1
- **semantic-release** configuration: default
- CI logs:  
https://circleci.com/gh/graphcool/chromeless/374

Could this have to do with the changes on `gitHead` resolution?
Since the npm publish was successful the next build passes now because it doesn't publish the GitHub release.

**Comment by pvdlg** [comment]
If you are using the npm plugin there is no change in behavior regarding the `gitHead`. The code just has been moved to the plugin.

The error happened when the github plugin create the new tag with `await github.gitdata.createReference({owner, repo, ref, sha: gitHead});`

The `gitHead` passed here is the sha of the last commit in the branch obtained with `await execa('git', ['rev-parse', 'HEAD']);`

It's probably a temporary problem with the GitHub API. Or it might be that the last commit sha in the local repo is not present in the remote (on GitHub).

You can obtain more info if you run `semantic-release` in debug mode with the `DEBUG` environment variable set to `sematic-release:*`.

**Comment by felixfbecker** [comment]
This was running in CI so the commit must have existed. I cannot reproduce anymore because now the version was already published to npm

**Comment by pvdlg** [comment]
We had several issue related to temporary errors with the GitHub API. Maybe it's on of this situation.
I really don't why GitHub would return a 404 when we create a tag that reference the last commit sha in the repo.

**Comment by gr2m** [comment]
I have seen random 404s from GitHub’s API as well. They way I worked around them is to implement  a 1s delay between write requests. see https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits

> If you're making a large number of POST, PATCH, PUT, or DELETE requests for a single user or client ID, wait at least one second between each request.

This is something that I plan to built into `@octokit/rest` core (renamed for `github`), so these errors should go away

**Comment by pvdlg** [comment]
Are the 404 due to go over the rate limit? Should it be another type of error like 429?

**Comment by felixfbecker** [comment]
I also think a retry logic would be good here, given how annoying it is if you ever get into the state were it's released to npm but not to github

**Comment by gr2m** [comment]
> Are the 404 due to go over the rate limit? Should it be another type of error like 429?

I know, there are tons of different abuse limit mechanisms. I thought it might be related to this

> Requests that create content which triggers notifications, such as issues, comments and pull requests, may be further limited and will not include a Retry-After header in the response. Please create this content at a reasonable pace to avoid further limiting

But in this case, creating a tag, does not trigger notifications, so that’s new to me, too

**Comment by pvdlg** [comment]
> But in this case, creating a tag, does not trigger notifications, so that’s new to me, too

I think creating a tag make a notification in the News feed of people who watch the repo (Or maybe it's on the actual release?). I don't know if that count as a notification per the API definition.

**Comment by gr2m** [comment]
I’ll investigate this with the GitHub platform team, I’m curious myself. 

Either way, adding the 1s delay for write operations will fix it. 

It might take a while until this lands in `@octokit/rest`, but I can create a `@semantic-release/octokit-rest` package that adds that for us using experimental APIs of the `@octokit/rest` package. I’ll have time next week Tuesday to look into it

**Comment by felixfbecker** [comment]
👎  on a `@semantic-release/octokit-rest` fork - timeouts should be easy enough to add in semantic-release and if there is experimental API to try out, that should not be semantic-release specific.

**Comment by gr2m** [comment]
not a fork, a module that depends on and pins the version of `@octokit/rest`. We would use the `octokit.hook` API (which is available, but not yet documented, hence experimental) which might or might not break between patch releases. That way can make our semantical versions

**Comment by gr2m** [comment]
I’m assuming here that we would use `@octokit/rest` in more than one of the semantic-release modules? If it’s just one, we can also put the logic in there and pin the version of `@octokit/rest`

**Comment by felixfbecker** [comment]
I don't know why it would be used in more than `@semantic-release/github`. All other modules should not be coupled to GitHub. Therefor imo another module sounds overly complicated.

**Comment by boennemann** [comment]
GitHub API returning returning random errors and requiring 1sec delay sounds awfully similar to the problems Greenkeeper runs into.

Referencing the solutions for the very same problems:

retry: https://github.com/greenkeeperio/greenkeeper/blob/e3b2ec236139c0202df840f33187d75deca9222e/lib/github.js#L6-L27

1sec delay: https://github.com/greenkeeperio/greenkeeper/blob/e3b2ec236139c0202df840f33187d75deca9222e/lib/github-queue.js

tl;dr `require('promise-retry')` / `require('promise-queue')`

**Comment by gr2m** [comment]
A wild @boennemann appears 🎉 👋  Thanks for the links :)

@pvdlg can you confirm that `@semantic-release/github` is the only one using `github`/`@octokit/rest` right now?

Either way, I can work on this next week. See the retry link above in Stephan’s comment to see how it’s currently implemented in Greenkeeper using the `octokit.hook` API. Here is a simple usage example of the hook api: https://github.com/octokit/rest.js/tree/master/lib/plugins. It’s https://github.com/gr2m/before-after-hook under the hood

**Comment by pvdlg** [comment]
I can confirm that `@octokit/rest` is used only in `@semantic-release/github`.

It seems the issue is related to the GitHub backend that for some reasons return random errors. I'm not sure the solution should be implemented in `semantic-release`. It should probably be handled on the GitHub backend (by return a proper 429) or in the API client. At least it worth reporting the problem to GitHub and see what they say. If the issue is urgent we can implement a retry based workaround in `@semantic-release/github`.

I don't think `octokit.hook` is the best solution though. `octokit.hook` allow to run something before or after. What we want to do here is to retry on error for a certain number of time, so we should implement a solution based on something like `promise-retry` as suggested by @boennemann. 

Actually we already have something similar done with [p-retry](https://github.com/sindresorhus/p-retry) in [test/helpers/npm-registry.js](https://github.com/semantic-release/semantic-release/blob/caribou/test/helpers/npm-registry.js#L36) and [travis-deploy-once/lib/wait-for-other-jobs.js](https://github.com/semantic-release/travis-deploy-once/blob/master/lib/wait-for-other-jobs.js). The advantage of `p-retry` is that we can use [exponential backoff](https://github.com/tim-kos/node-retry#retrytimeoutsoptions).

We could implement a similar solutions for the calls we expect to not fail (like creating a tag or a release). We could do something like:
- retries: 3
- factor: 2
- minTimeout: 1000

So if the first call fails we would wait 1s and retry, if it still fails wait 2s and retry, and if it still fails wait 4s and retry.

Another thing to consider is that several times when I experienced these random errors, they would persist for a couple hours. I was retrying the calls but still obtained the same error. After waiting a couple hours it would work right away. For those type of situation doing a retry in the code would not work.

**Comment by gr2m** [comment]
> I don't think octokit.hook is the best solution though. octokit.hook allow to run something before or after

Yes it’s not a very elegant solution for the retry, I’m thinking about that right now and will work on it next week

> It should probably be handled on the GitHub backend (by return a proper 429) or in the API client. At least it worth reporting the problem to GitHub and see what they say

Yes I’m talking to them, I’ll keep you posted.

> Another thing to consider is that several times when I experienced these random errors, they would persist for a couple hours

Locally, or on CI? I’ve seen it on travis in a cron job that always runs at the same time. After adding the 1s delay, I haven’t seen it happening again

**Comment by pvdlg** [comment]
> Yes it’s not a very elegant solution for the retry, I’m thinking about that right now and will work on it next week

👍I guess it worth having the GitHub feedback, to decide if it's better done in `semantic-release` as a temporary workaround or permanently in `@octokit/rest`.

For info it's implemented in the npm registry client: https://github.com/npm/npm-registry-client/blob/80870c565700699111a56bb1b76f28202442aba5/lib/request.js#L67. It's a common thing for API to implement such feature.

The tough part is to determine when to retry and when to consider the error and actual error that doesn't need a retry.

> Locally, or on CI?

Both. In order to debug a problem that happened on the CI on one of my repo I was trying to make the call locally with [Postman](https://www.getpostman.com/). Couldn't figure out why I was having the error (a 404 as well if I remember well) so I gave up. I went back to it a few hours later, same request worked right away.
